File: src/main/java/net/flyingfishflash/loremlist/domain/lrmitem/LrmItemServiceDefault.java
package net.flyingfishflash.loremlist.domain.lrmitem;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.ConstraintViolationException;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.function.IntSupplier;
import net.flyingfishflash.loremlist.core.exceptions.CoreException;
import net.flyingfishflash.loremlist.domain.ServiceResponse;
import net.flyingfishflash.loremlist.domain.exceptions.DomainException;
import net.flyingfishflash.loremlist.domain.lrmitem.data.LrmItemDeleted;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmList;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListService;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct;
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItemService;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
public class LrmItemServiceDefault implements LrmItemService {

  private final LrmItemRepository lrmItemRepository;
  private final LrmListService lrmListService;
  private final LrmListItemService lrmListItemService;
  private final ObjectMapper objectMapper;
  private final Validator validator = Validation.buildDefaultValidatorFactory().getValidator();

  public LrmItemServiceDefault(
      LrmItemRepository lrmItemRepository,
      LrmListService lrmListService,
      LrmListItemService lrmListItemService,
      ObjectMapper objectMapper) {
    this.lrmItemRepository = lrmItemRepository;
    this.lrmListService = lrmListService;
    this.lrmListItemService = lrmListItemService;
    this.objectMapper = objectMapper;
  }

  @Override
  public ServiceResponse<Long> countByOwner(String owner) {
    try {
      long repositoryResponse = lrmItemRepository.countByOwner(owner);
      String message =
          repositoryResponse == 1L ? repositoryResponse + " item." : repositoryResponse + " items.";
      return new ServiceResponse<>(repositoryResponse, message);
    } catch (Exception cause) {
      throw DomainException.builder()
          .cause(cause)
          .message("Total item count couldn't be generated.")
          .build();
    }
  }

  @Override
  public ServiceResponse<LrmItemDeleted> deleteByOwner(String owner) {
    try {
      List<LrmItem> items = findByOwner(owner).getContent();
      if (!items.isEmpty()) {
        for (LrmItem item : items) {
          if (!item.lists().isEmpty()) {
            lrmListItemService.removeByOwnerAndItemId(item.id(), owner);
          }
        }
        Set<UUID> ids =
            items.stream().map(LrmItem::id).collect(java.util.stream.Collectors.toSet());
        lrmItemRepository.deleteById(ids);
      }
      List<String> itemNames = items.stream().map(LrmItem::name).sorted().toList();
      List<String> associatedListNames =
          items.stream()
              .flatMap(item -> item.lists().stream())
              .map(LrmListSuccinct::getName)
              .sorted()
              .toList();
      LrmItemDeleted lrmItemDeleted = new LrmItemDeleted(itemNames, associatedListNames);
      return new ServiceResponse<>(
          lrmItemDeleted,
          "Deleted all ("
              + lrmItemDeleted.itemNames().size()
              + ") of your items, and removed them from "
              + lrmItemDeleted.associatedListNames().size()
              + " lists.");
    } catch (Exception cause) {
      if (cause instanceof CoreException coreException) {
        throw DomainException.builder()
            .cause(coreException)
            .httpStatus(coreException.getHttpStatus())
            .message("No items were deleted: " + coreException.getResponseMessage())
            .supplemental(coreException.getSupplemental())
            .build();
      }
      throw DomainException.builder().cause(cause).message("No items were deleted.").build();
    }
  }

  @Override
  public ServiceResponse<LrmItemDeleted> deleteByOwnerAndId(
      UUID id, String owner, boolean removeListAssociations) {
    try {
      LrmItem item = findByOwnerAndId(id, owner).getContent();
      LrmItemDeleted lrmItemDeleteResponse = createDeleteResponse(item);
      if (!lrmItemDeleteResponse.associatedListNames().isEmpty()) {
        if (removeListAssociations) {
          lrmListItemService.removeByOwnerAndItemId(id, owner);
        } else {
          throwListAssociationException(lrmItemDeleteResponse);
        }
      }
      doDeleteByOwnerAndId(id, owner);
      String noun = lrmItemDeleteResponse.associatedListNames().size() == 1 ? "list" : "lists";
      return new ServiceResponse<>(
          lrmItemDeleteResponse,
          "Deleted item '"
              + lrmItemDeleteResponse.itemNames().get(0)
              + "', and removed it from "
              + lrmItemDeleteResponse.associatedListNames().size()
              + " "
              + noun
              + ".");
    } catch (Exception cause) {
      if (cause instanceof CoreException coreException) {
        throw DomainException.builder()
            .cause(coreException)
            .httpStatus(coreException.getHttpStatus())
            .message(
                "Item id " + id + " could not be deleted: " + coreException.getResponseMessage())
            .supplemental(coreException.getSupplemental())
            .build();
      }
      throw DomainException.builder()
          .cause(cause)
          .message("Item id " + id + " could not be deleted.")
          .build();
    }
  }

  private LrmItemDeleted createDeleteResponse(LrmItem item) {
    List<String> associatedListNames =
        item.lists().stream().map(LrmListSuccinct::getName).sorted().toList();
    return new LrmItemDeleted(List.of(item.name()), associatedListNames);
  }

  private void throwListAssociationException(LrmItemDeleted lrmItemDeleteResponse) {
    String message =
        "Item '"
            + lrmItemDeleteResponse.itemNames().get(0)
            + "' is associated with "
            + lrmItemDeleteResponse.associatedListNames().size()
            + " list(s). First remove the item from each list.";
    throw DomainException.builder()
        .httpStatus(HttpStatus.UNPROCESSABLE_ENTITY)
        .supplemental(
            Map.of(
                "itemNames", objectMapper.valueToTree(lrmItemDeleteResponse.itemNames()),
                "associatedListNames",
                    objectMapper.valueToTree(lrmItemDeleteResponse.associatedListNames())))
        .message(message)
        .build();
  }

  private void doDeleteByOwnerAndId(UUID itemId, String owner) {
    int deletedCount = lrmItemRepository.deleteByOwnerAndId(itemId, owner);
    if (deletedCount != 1) {
      handleInvalidAffectedRecordCount(deletedCount, itemId);
    }
  }

  @Override
  public ServiceResponse<List<LrmItem>> findByOwner(String owner) {
    try {
      List<LrmItem> repositoryResponse = lrmItemRepository.findByOwner(owner);
      return new ServiceResponse<>(
          repositoryResponse, "Retrieved all items owned by " + owner + ".");
    } catch (Exception cause) {
      throw DomainException.builder().cause(cause).message("Items could not be retrieved.").build();
    }
  }

  @Override
  public ServiceResponse<LrmItem> findByOwnerAndId(UUID id, String owner) {
    LrmItem repositoryResponse;
    try {
      repositoryResponse = lrmItemRepository.findByOwnerAndIdOrNull(id, owner);
    } catch (Exception cause) {
      throw DomainException.builder()
          .cause(cause)
          .message("Item id " + id + " could not be retrieved.")
          .build();
    }
    if (repositoryResponse == null) {
      throw new ItemNotFoundException(id);
    }
    return new ServiceResponse<>(
        repositoryResponse, "Retrieved item '" + repositoryResponse.name() + "'");
  }

  @Override
  public ServiceResponse<List<LrmItem>> findByOwnerAndHavingNoListAssociations(String owner) {
    try {
      List<LrmItem> repositoryResponse =
          lrmItemRepository.findByOwnerAndHavingNoListAssociations(owner);
      return new ServiceResponse<>(
          repositoryResponse,
          "Retrieved " + repositoryResponse.size() + " items that are not a part of a list.");
    } catch (Exception cause) {
      throw DomainException.builder()
          .cause(cause)
          .message("Items without list associations could not be retrieved.")
          .build();
    }
  }

  @Override
  public ServiceResponse<List<LrmItem>> findByOwnerAndHavingNoListAssociations(
      String owner, UUID listId) {
    String exceptionMessage =
        "Items eligible to be added to list '" + listId + "' could not be retrieved";
    HttpStatus httpStatus = null;
    try {
      LrmList lrmList = lrmListService.findByOwnerAndId(listId, owner).getContent();
      List<LrmItem> repositoryResponse =
          lrmItemRepository.findByOwnerAndHavingNoListAssociations(owner, listId);
      return new ServiceResponse<>(
          repositoryResponse,
          "Retrieved "
              + repositoryResponse.size()
              + " items eligible to be added to list '"
              + lrmList.name()
              + "'.");
    } catch (Exception cause) {
      if (cause instanceof DomainException domainException) {
        httpStatus = domainException.getHttpStatus();
      }
      String message = exceptionMessage + ": " + cause.getMessage();
      throw DomainException.builder().cause(cause).message(message).httpStatus(httpStatus).build();
    }
  }

  @Override
  public void patchName(LrmItem patchedLrmItem) {
    patchItem(patchedLrmItem, () -> lrmItemRepository.updateName(patchedLrmItem));
  }

  @Override
  public void patchDescription(LrmItem patchedLrmItem) {
    patchItem(patchedLrmItem, () -> lrmItemRepository.updateDescription(patchedLrmItem));
  }

  private void patchItem(LrmItem patchedLrmItem, IntSupplier updateAction) {
    validateItemEntity(patchedLrmItem);
    int updatedCount = updateAction.getAsInt();
    if (updatedCount != 1) {
      handleInvalidAffectedRecordCount(updatedCount, patchedLrmItem.id());
    }
  }

  private void handleInvalidAffectedRecordCount(int affectedRecordCount, UUID id) {
    if (affectedRecordCount < 1) {
      throw DomainException.builder()
          .message("No item affected by the repository operation.")
          .build();
    } else if (affectedRecordCount > 1) {
      throw DomainException.builder()
          .message("More than one item with id " + id + " were found.")
          .build();
    } else {
      throw new IllegalArgumentException(
          affectedRecordCount + " should not be passed to this function");
    }
  }

  private void validateItemEntity(LrmItem lrmItem) {
    Set<ConstraintViolation<LrmItem>> violations = validator.validate(lrmItem);
    if (!violations.isEmpty()) {
      throw new ConstraintViolationException(violations);
    }
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/domain/lrmitem/LrmItemServiceDefault.kt
package net.flyingfishflash.loremlist.domain.lrmitem

import jakarta.validation.ConstraintViolationException
import jakarta.validation.Validation
import net.flyingfishflash.loremlist.core.exceptions.CoreException
import net.flyingfishflash.loremlist.domain.ServiceResponse
import net.flyingfishflash.loremlist.domain.exceptions.DomainException
import net.flyingfishflash.loremlist.domain.lrmitem.data.LrmItemDeleted
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItemService
import net.flyingfishflash.loremlist.toJsonElement
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
@Transactional
class LrmItemServiceDefault(private val lrmItemRepository: LrmItemRepository, private val lrmListItemService: LrmListItemService) : LrmItemService {
  private val validator = Validation.buildDefaultValidatorFactory().validator

  override fun countByOwner(owner: String): ServiceResponse<Long> {
    return runCatching {
      val repositoryResponse = lrmItemRepository.countByOwner(owner)
      val message = if (repositoryResponse == 1L) "$repositoryResponse item." else "$repositoryResponse items."
      return@runCatching ServiceResponse(content = repositoryResponse, message = message)
    }.getOrElse { cause ->
      throw DomainException(cause = cause, message = "Total item count couldn't be generated.")
    }
  }

  override fun deleteByOwner(owner: String): ServiceResponse<LrmItemDeleted> {
    return runCatching {
      val items = findByOwner(owner = owner).content
      if (items.isNotEmpty()) {
        items.filter { it.lists.isNotEmpty() }.forEach {
          lrmListItemService.removeByOwnerAndItemId(itemId = it.id, owner = owner)
        }
        lrmItemRepository.deleteById(ids = items.map { it.id }.toSet())
      }
      val lrmItemDeleted = LrmItemDeleted(
        itemNames = items.map { it.name }.sorted(),
        associatedListNames = items.flatMap { it.lists }.map { it.name }.sorted(),
      )
      return@runCatching ServiceResponse(
        content = lrmItemDeleted,
        message = "Deleted all (${lrmItemDeleted.itemNames.size}) of your items, and removed them from ${lrmItemDeleted.associatedListNames.size} lists.",
      )
    }.getOrElse { cause ->
      when (cause) {
        is CoreException -> throw DomainException(
          cause = cause,
          httpStatus = cause.httpStatus,
          message = "No items were deleted: ${cause.responseMessage}",
          supplemental = cause.supplemental,
        )
        else -> throw DomainException(cause = cause, message = "No items were deleted.")
      }
    }
  }

  override fun deleteByOwnerAndId(id: UUID, owner: String, removeListAssociations: Boolean): ServiceResponse<LrmItemDeleted> {
    return runCatching {
      val item = findByOwnerAndId(id = id, owner = owner).content
      val lrmItemDeleteResponse = createDeleteResponse(item)
      if (lrmItemDeleteResponse.associatedListNames.isNotEmpty()) {
        if (removeListAssociations) {
          lrmListItemService.removeByOwnerAndItemId(itemId = id, owner = owner)
        } else {
          throwListAssociationException(lrmItemDeleteResponse)
        }
      }
      doDeleteByOwnerAndId(itemId = id, owner = owner)
      val noun = if (lrmItemDeleteResponse.associatedListNames.size == 1) "list" else "lists"
      return@runCatching ServiceResponse(
        content = lrmItemDeleteResponse,
        message = "Deleted item '${lrmItemDeleteResponse.itemNames.first()}', and removed it from ${lrmItemDeleteResponse.associatedListNames.size} $noun.",
      )
    }.getOrElse { cause ->
      when (cause) {
        is CoreException -> throw DomainException(
          cause = cause,
          httpStatus = cause.httpStatus,
          message = "Item id $id could not be deleted: ${cause.responseMessage}",
          supplemental = cause.supplemental,
        )
        else -> throw DomainException(cause = cause, message = "Item id $id could not be deleted.")
      }
    }
  }

  private fun createDeleteResponse(item: LrmItem): LrmItemDeleted {
    return LrmItemDeleted(
      itemNames = listOf(item.name),
      associatedListNames = item.lists.map { it.name }.sorted(),
    )
  }

  private fun throwListAssociationException(lrmItemDeleteResponse: LrmItemDeleted) {
    val message = "Item '${lrmItemDeleteResponse.itemNames.first()}' " +
      "is associated with ${lrmItemDeleteResponse.associatedListNames.size} list(s). First remove the item from each list."
    throw DomainException(
      httpStatus = HttpStatus.UNPROCESSABLE_ENTITY,
      supplemental = mapOf(
        "itemNames" to lrmItemDeleteResponse.itemNames.toJsonElement(),
        "associatedListNames" to lrmItemDeleteResponse.associatedListNames.toJsonElement(),
      ),
      message = message,
    )
  }

  private fun doDeleteByOwnerAndId(itemId: UUID, owner: String) {
    val deletedCount = lrmItemRepository.deleteByOwnerAndId(id = itemId, owner = owner)
    if (deletedCount != 1) {
      handleInvalidAffectedRecordCount(affectedRecordCount = deletedCount, id = itemId)
    }
  }

  override fun findByOwner(owner: String): ServiceResponse<List<LrmItem>> {
    return runCatching {
      val repositoryResponse = lrmItemRepository.findByOwner(owner = owner)
      return@runCatching ServiceResponse(content = repositoryResponse, message = "Retrieved all items owned by $owner.")
    }.getOrElse { cause ->
      throw DomainException(cause = cause, message = "Items could not be retrieved.")
    }
  }

  override fun findByOwnerAndId(id: UUID, owner: String): ServiceResponse<LrmItem> {
    val repositoryResponse = runCatching {
      return@runCatching lrmItemRepository.findByOwnerAndIdOrNull(id = id, owner = owner)
    }.getOrElse { cause ->
      throw DomainException(cause = cause, message = "Item id $id could not be retrieved.")
    } ?: throw ItemNotFoundException(id = id)
    return ServiceResponse(content = repositoryResponse, message = "Retrieved item '${repositoryResponse.name}'")
  }

  override fun findByOwnerAndHavingNoListAssociations(owner: String): ServiceResponse<List<LrmItem>> {
    val exceptionMessage = "Items without list associations could not be retrieved."
    return runCatching {
      val repositoryResponse = lrmItemRepository.findByOwnerAndHavingNoListAssociations(owner = owner)
      return@runCatching ServiceResponse(content = repositoryResponse, message = "Retrieved ${repositoryResponse.size} items that are not a part of a list.")
    }.getOrElse { cause ->
      throw DomainException(cause = cause, message = exceptionMessage)
    }
  }

  // TODO: Wrap in runCatching() and ServiceResponse
  override fun patchName(patchedLrmItem: LrmItem) {
    patchItem(patchedLrmItem) { lrmItemRepository.updateName(patchedLrmItem) }
  }

  override fun patchDescription(patchedLrmItem: LrmItem) {
    patchItem(patchedLrmItem) { lrmItemRepository.updateDescription(patchedLrmItem) }
  }

  private fun patchItem(patchedLrmItem: LrmItem, updateAction: () -> Int) {
    validateItemEntity(patchedLrmItem)
    val updatedCount = updateAction()
    if (updatedCount != 1) {
      handleInvalidAffectedRecordCount(updatedCount, patchedLrmItem.id)
    }
  }

  private fun handleInvalidAffectedRecordCount(affectedRecordCount: Int, id: UUID) {
    when {
      affectedRecordCount < 1 -> throw DomainException(message = "No item affected by the repository operation.")
      affectedRecordCount > 1 -> throw DomainException(message = "More than one item with id $id were found.")
      else -> throw IllegalArgumentException("$affectedRecordCount should not be passed to this function")
    }
  }

  private fun validateItemEntity(lrmItem: LrmItem) {
    val violations = validator.validate(lrmItem)
    if (violations.isNotEmpty()) {
      throw ConstraintViolationException(violations)
    }
  }
}
